### ISSUE TITLE
Migrate existing text extensions to the new structure

### ISSUE THREAD

**Issue by ValWood** [pombase/canto#1015]
transferred from canto/issues/1012

Kim:
In the new annotation extension code extensions are stored in a structured way. We need to parse the existing free text extensions and report anything that isn't allowed by the config files.

Need to decide what to do with "qualifier=different_pathway", "qualifier=contributes_to" and "qualifier=NOT". "col17=PR:000027576"

Here are the current warnings:
https://www.dropbox.com/s/a3j9qvgezi1z9xs/v10_extension_warnings.txt?dl=0

Many of them (like "qualifier=NOT") will be fixed by the upgrade script. Let me know which of the qualifiers in the extensions should be kept and I'll remove the warnings for them.

**Comment by kimrutherford** [comment]
There is no way in the new interface to add a NOT or col17 qualifier.  We need to decide what to do about that.

**Comment by ValWood** [comment]
In the meantime curators can fix some

Many others already get caught by the logs.
We also need any new types, not already recorded in the logs,  to be recorded there too

**Comment by Antonialock** [comment]
I have fixed all that I can fix. the rest are residue=, and your reminders to go back and update stuff val

**Comment by ValWood** [comment]
Thanks,

Kim can you regenerate file after next chado load?]Then we can adress the rest.

Val

**Comment by kimrutherford** [comment]
Here are the current warnings when running the upgrade script:
  https://www.dropbox.com/s/0qhq7p0ng07cz1s/upgrade-to-v10-extension.log?dl=0

**Comment by ValWood** [comment]
I'm fixing these, which are mainly in sessions I shouldn't have put through approval yet:

failed to store extension in 7d8eddd77bc43a59: upgrade script can't handle: exists_during=(GO:0051329)
failed to store extension in b42a7324f08b6a24: upgrade script can't handle: qualifier=inhibits
failed to store extension in 670e6db747c0dacb: can't parse 'this should be a child of other terms  add  multiseptate swollen  same for cdc16-116mutant to the restrictive temperature (Minetetal.,1979)', in: this should be a child of other terms  add  multiseptate swollen  same for cdc16-116mutant to the restrictive temperature (Minetetal.,1979)
failed to store extension in 47e7d9844090c3e4: can't parse '(APC-Ste9 dependent protein destruction/11365/)', in: has_input(PomBase:cdc13)|has_input(PomBase:cig1)|happens_during(GO:0000080)| (APC-Ste9 dependent protein destruction/11365/)
failed to store extension in 7d2c05ef7277f909: can't parse 'AT KINETOCHORE BOLISHDD', in: AT KINETOCHORE BOLISHDD
failed to store extension in 7d2c05ef7277f909: can't parse 'arms', in: arms
failed to store extension in 7d2c05ef7277f909: can't parse 'kinetochore', in: kinetochore
failed to store extension in 7d2c05ef7277f909: can't parse 'kinetochore', in: kinetochore
failed to store extension in 7d2c05ef7277f909: can't parse 'condensin', in: condensin, which subunit assayed? has_expressivity(FYPO_EXT:0000001)
failed to store extension in 7d2c05ef7277f909: can't parse 'condensin', in: condensin, which subunit assayed?
failed to store extension in 7d2c05ef7277f909: can't parse 'condensin', in: condensin, which subunit assayed?
failed to store extension in 7d2c05ef7277f909: can't parse 'ARMS assayed_using(cnd2)', in: ARMS assayed_using(cnd2)
failed to store extension in 80821af8a094ff52: can't parse 'requested more specific term during mitotic interphase', in: requested more specific term during mitotic interphase
failed to store extension in a42509fe62ac69ca: can't parse 'reduced', in: reduced
failed to store extension in a42509fe62ac69ca: can't parse 'reduced', in: reduced
failed to store extension in a42509fe62ac69ca: can't parse 'M-phase', in: M-phase
failed to store extension in a42509fe62ac69ca: can't parse 'M-phase', in: M-phase
failed to store extension in 1f8a9a6747848375: upgrade script can't handle: qualifier=same_pathway
failed to store extension in 3ddc3ca948411f8e: can't parse 'H4 K5 K8 K12', in: H4 K5 K8 K12
failed to store extension in 01c913b208de3dfb: upgrade script can't handle: qualifier=predominantly
failed to store extension in 8d860f7ec28a89b2: can't parse 'DURING interphase assayed_using(PomBase:cut3)', in: DURING interphase assayed_using(PomBase:cut3)

**Comment by kimrutherford** [comment]
> I'm fixing these, which are mainly in sessions I shouldn't have put through approval yet:

The log output I sent was from *all* sessions as they all need to have their extension strings migrated to the sane new world.

**Comment by ValWood** [comment]
Yes, this will stop me from using the extensions field for my notes. I don't put the sessions through approval anymore but I still use it. I shouldn't...

**Comment by ValWood** [comment]
fixed most of the above. 
Can you regenerate the file? we can discuss the remaining qualifiers today

**Comment by kimrutherford** [comment]
OK, I'm re-running it now.  Takes a few minutes.

**Comment by kimrutherford** [comment]
> OK, I'm re-running it now. Takes a few minutes.

Sorry, I forgot this last night.  Here's the new log:
  https://www.dropbox.com/s/v7qokrhayqnjkhn/upgrade-to-v10-extension-2.log?dl=0

**Comment by ValWood** [comment]
OK all now fixed except
failed to store extension in 0b3d460aeaeeeb91: upgrade script can't handle: qualifier=heterologous_system
failed to store extension in 0b3d460aeaeeeb91: upgrade script can't handle: qualifier=heterologous_system
failed to store extension in 025da8633ca22521: upgrade script can't handle: qualifier=different_pathway
failed to store extension in 025da8633ca22521: upgrade script can't handle: qualifier=different_pathway
failed to store extension in 134ecf3bebd24e1a: upgrade script can't handle: qualifier=major
failed to store extension in 134ecf3bebd24e1a: upgrade script can't handle: qualifier=minor
failed to store extension in 134ecf3bebd24e1a: upgrade script can't handle: qualifier=major
failed to store extension in 7f9fcf3e0233cf42: upgrade script can't handle: qualifier=major
failed to store extension in dde3c6ac603a05eb: upgrade script can't handle: qualifier=major
failed to store extension in 1f8a9a6747848375: upgrade script can't handle: qualifier=same_pathway
failed to store extension in 01c913b208de3dfb: upgrade script can't handle: qualifier=predominantly

heterologous_system should be moved to artemis
the rest should probably be moved to "comment"  we  need to decide if we want to keep these (we have more in Artemis). They are useful but is it worth the effort.

This will leave the ones that we can't ignore. 
residue= (will be sorted)
contributes_to 
NOT 
colocalizes_with 
PR
We need to decide  how to handle  these....soon

**Comment by ValWood** [comment]
Kim there might be a few more. At the front of this file there are a load of errors like this:

Use of uninitialized value $_ in pattern match (m//) at lib/Canto/Config/ExtensionConf.pm line 69, <$conf_fh> line 7.
Use of uninitialized value $_ in pattern match (m//) at lib/Canto/Config/ExtensionConf.pm line 69, <$conf_fh> line 8.
Use of uninitialized value $_ in pattern match (m//) at lib/Canto/Config/ExtensionConf.pm line 69, <$conf_fh> line 9.
Use of uninitialized value $_ in pattern match (m//) at lib/Canto/Config/ExtensionConf.pm line 69, <$conf_fh> line 10.
Use of uninitialized value $_ in pattern match (m//) at lib/Canto/Config/ExtensionConf.pm line 69, <$conf_fh> line 11.
Use of uninitialized value $_ in pattern match (m//) at lib/Canto/Config/ExtensionConf.pm line 69, <$conf_fh> line 12.
Use of uninitialized value $_ in pattern match (m//) at lib/Canto/Config/ExtensionConf.pm line 69, <$conf_fh> line 13.
Use of uninitialized value $_ in pattern match (m//) at lib/Canto/Config/ExtensionConf.pm line 69, <$conf_fh> line 14.
Use of uninitialized value $_ in pattern match (m//) at lib/Canto/Config/ExtensionConf.pm line 69, <$conf_fh> line 15.
Use of uninitialized value $_ in pattern match (m//) at lib/Canto/Config/ExtensionConf.pm line 69, <$conf_fh> line 16.
Use of uninitialized value $_ in pattern match (m//) at lib/Canto/Config/ExtensionConf.pm line 69, <$conf_fh> line 17.
Use of uninitialized value $_ in pattern match (m//) at lib/Canto/Config/ExtensionConf.pm line 69, <$conf_fh> line 18.
Use of uninitialized value $_ in pattern match (m//) at lib/Canto/Config/ExtensionConf.pm line 69, <$conf_fh> line 19.

**Comment by kimrutherford** [comment]
Sorry about the error messages.  I don't think you will have missed anything.  Here is the latest log, with the code fixed:
  https://www.dropbox.com/s/jun1wazui864g46/upgrade-to-v10-extension-3.log?dl=0

**Comment by ValWood** [comment]
In the fist instance these could be "admin only". 
https://github.com/pombase/canto/issues/1011
I doubt the community will ever use them....

**Comment by kimrutherford** [comment]
Val from #985:

> Any ideas for the GO qualifiers? Can we treat these int he same way as residue=?

Which ones are the GO qualifiers?  I guess "col17=" is one?  It's not a relation and it gets stored a bit differently in Canto but we could use the extension interface to add/edit them.

For example, you could add this extension configuration:

     GO:0005575   is_a  column_17  Text   Gene product form ID (column 17)   0,1   admin

so that admins would be prompted for a product.

Here's the full list of qualifier values and the number of times the value is used:

     14 colocalizes_with
     56 contributes_to
      2 different_pathway
      4 major
      1 minor
     42 NOT
      1 predominantly
      1 same_pathway

These mostly look like flags.  I can special case them just to get them into Chado but you won't be able to add more in future.  Do we need a (maybe admin only) way to add flags to an annotation from a small controlled vocabulary?  If that's the case could you make a ticket or tickets?  It might be possible to add it in a hacky way to the extension interface.

**Comment by kimrutherford** [comment]
I've changed the loader so that col17= and column_17= extensions are stored in the new Canto as if they are relations: "column_17(...)"

That's got rid of a few warnings.

Here's the latest log:
  https://www.dropbox.com/s/hiqo3ztymcxdy97/upgrade-to-v10-extension-7.log?dl=0

**Comment by mah11** [comment]
I'm not thrilled about the prospect of muddying the distinction between column 17, qualifiers (GAF column 4), and extensions (GAF column 16), because they have different meanings. A col 17 entry for a GO annotation is more analogous to an allele in a phenotype annotation than to an extension. GO qualifiers don't really have an equivalent/analog in other annotation types.

But I guess if the extension interface is the least-worst thing to hack for them, I'll try to only grumble a little bit.

**Comment by ValWood** [comment]
So, they represent different things, but it is my understanding that Chado (and Ensembl) have one way to store this 'associated data' . I'm not so bothered that they are treated as 'relations'...I think they should be in a different table though, is that the case? 
i.e a table for extension, a table for qualifiers and a table for  "Gene Product Form" (col 17)

**Comment by mah11** [comment]
I think it'll help users grasp the various meanings if the Canto interface doesn't lump these, even if the db schemas do. I know the qualifiers and col17 will be admin-only for a while, but probably not forever, and it won't hurt for us to have it laid out sensibly anyway.

When you say "think they should be in a different table", do you mean in the database schema? I don't even know the Chado schema, let alone the Ensembl schema, well enough to say where each of these bits goes. (Tangent: in the GO db schema, I think qualifiers and extensions go in different tables, and I don't know where col17 gp forms go, but it doesn't look like it's either of the tables for quals or exts.)

**Comment by kimrutherford** [comment]
Midori and I talked about this on Skype.

For col17, we think that the best we can to configure the extension interface with text like "isoform (optional)" and then have a free text box.  We can't expect the community to request PR:... terms.

> I doubt the community will ever use them....

We think so too.  For the other qualifiers we think it's best to encourage the users to put the details in the comment box.  The qualifiers aren't that common so showing them for every annotation will clutter the interface and may be confusing.  But we think admin users should be able to add these qualifiers on the extension page.  I'll work on that (#1046).

"residue=" is already covered - we'll just treat them as extensions.

I think with just need to add the appropriate configuration for the col17 qualifiers and then we can close this.

**Comment by kimrutherford** [comment]
We're down to just a few warnings when fixing up free text extension strings:
  https://www.dropbox.com/s/agrpejg8q0vog0n/qualifier_warnings_upgrade_to_v10-2015-12-16.txt?dl=0

I can easily store qualifier=major, qualifier=different_pathway etc. in the same way as the GO qualifiers if you think that's a good idea.

**Comment by ValWood** [comment]
Lets leave them in the logs for now. Likely we will get rid of them, but we might decide we need them sometimes (even as a stop gap).

Does this file output now appear in the logs directory?

(Lots of the errors in here are in my in progress sessions, will hopefully be fixed today or tomorrow)

**Comment by kimrutherford** [comment]
> Lets leave them in the logs for now.

These are messages from when I test upgrading a copy of the live Canto database to the new extension interface code.  A warning means that some data will be discarded when the new code goes live, which will happen only once.

**Comment by ValWood** [comment]
OK decision.
I'm fixing mine today.

Any remaining qualifiers I  will store as comments for now

**Comment by kimrutherford** [comment]
There aren't many warnings left now.  Are there any of these that need fixing?  Any unfixed warnings will be turned into comment when the annotation extension changes go live.

    025da8633ca22521: failed to store qualifier with value 'different_pathway' in: qualifier=different_pathway
    025da8633ca22521: failed to store qualifier with value 'different_pathway' in: qualifier=different_pathway
    65edebcc485f98a8: can't parse 'residue((S12,S122,S264,S345,S418))', in: residue=(S12,S122,S264,S345,S418)
    134ecf3bebd24e1a: failed to store qualifier with value 'major' in: qualifier=major
    134ecf3bebd24e1a: failed to store qualifier with value 'minor' in: qualifier=minor
    134ecf3bebd24e1a: failed to store qualifier with value 'major' in: qualifier=major
    7f9fcf3e0233cf42: failed to store qualifier with value 'major' in: qualifier=major
    dde3c6ac603a05eb: failed to store qualifier with value 'major' in: qualifier=major
    1f8a9a6747848375: failed to store qualifier with value 'same_pathway' in: qualifier=same_pathway
    01c913b208de3dfb: failed to store qualifier with value 'predominantly' in: qualifier=predominantly
    558b22ef42aa6ec2: can't parse 'ABOLISHED tetermerization', in: ABOLISHED tetermerization
    558b22ef42aa6ec2: can't parse 'ABOLISHED tetermerization', in: ABOLISHED tetermerization
    558b22ef42aa6ec2: can't parse 'ABOLISHED tetermerization', in: ABOLISHED tetermerization
    dd0dc92db12b47e2: can't parse 'during mitotic interphase', in: during mitotic interphase
    dd0dc92db12b47e2: can't parse 'during mitotic M-phase', in: during mitotic M-phase
    dd0dc92db12b47e2: can't parse 'during mitotic M-phase', in: during mitotic M-phase
    dd0dc92db12b47e2: can't parse 'during mitotic M-phase', in: during mitotic M-phase
    6960ce3c7a8f53d0: can't parse 'AWAITING THESE SO TERMS: coincident_with(M and P silent
        cassettes)', in: AWAITING THESE SO TERMS: coincident_with(M and P silent
        cassettes), occurs_in(CL:0002674) | coincident_with(M cassette), occurs_in(CL:0002675)

**Comment by ValWood** [comment]
fixed
dd0dc92db12b47e2: can't parse 'during mitotic interphase', in: during mitotic interphase
dd0dc92db12b47e2: can't parse 'during mitotic M-phase', in: during mitotic M-phase
dd0dc92db12b47e2: can't parse 'during mitotic M-phase', in: during mitotic M-phase
dd0dc92db12b47e2: can't parse 'during mitotic M-phase', in: during mitotic M-phase

**Comment by ValWood** [comment]
Maybe we should move the others to comments pending decisions?

Not sure whats wrong with
65edebcc485f98a8: can't parse 'residue((S12,S122,S264,S345,S418))', in: residue=(S12,S122,S264,S345,S418)

whats the syntax for this one?

**Comment by Antonialock** [comment]
remove the bracket?

**Comment by ValWood** [comment]
Summary.
Kim now stores NOT,contributes to and colocalizes with as "has_qualifier"
column17 as a relation "column17" 
residue=

Questions (val)
Question, how do we curate these when the free text annotation_extension box has gone ?
Will this be an admin only option?

I migrated the remaining ones above to comments so the canto data will load into chado.

We can make things like things like "predominantly" available as a qualifier option if we decide that they are useful

**Comment by mah11** [comment]
I would very much like to have the extension box remain as an admin-only feature, because it would be a RRPITA to add more than a couple of extensions per annotation via the community-friendly interface ... and I've done papers that have had MANY extensions per annotation.

**Comment by kimrutherford** [comment]
> I would very much like to have the extension box remain as an admin-only feature

That could be done with a bit of re-engineering.  It would also mean you'd have continue to check and fix syntax problems reported in the log file.

**Comment by mah11** [comment]
>  It would also mean you'd have continue to check and fix syntax problems reported in the log file.

I could live with that. It'll be a lot less work than going through the step-by-step interface for the big sets of extensions (e.g. 073951bf43ce4011).

**Comment by ValWood** [comment]
To me it seems easier to duplicate the row and edit the extension? In this example the extension is nearly always the same so it wouldn't even usually need editing? Am I missing something?

I have no real objection to keeping it, I'm just wondering why its so much quicker this way?

**Comment by mah11** [comment]
If there are two or three extensions, then yes, it is almost as fast, and less error-prone, to duplicate and edit. But I have curated more than one or two papers where one annotation needs a dozen or more extensions. For those it would be prohibitively time-consuming and mouse-intensive to add them manually.

You will be free to ignore the text box anyway ;)

**Comment by kimrutherford** [comment]
I'll have a think about how best to implement this.  I like the idea of instant feed back that would let you know when the syntax and relations were correct.  I'd prefer if only valid annotation extension strings were stored in the Canto database because it makes loading into Chado more straight forward.

Midori, could you make a separate issue for idea?

**Comment by mah11** [comment]
#1063 opened

**Comment by ValWood** [comment]
Yeah and we do still need the text box for other stuff.

**Comment by kimrutherford** [comment]
> Yeah and we do still need the text box for other stuff.

What stuff are you thinking of?  I'd rather column17 and residue were treated like (admin-only) relations like we do for qualifiers (NOT, colocalizes_with etc).  So we'd have two extra links "column17 ..." and "residue ..." that just popup a text box - we would configure the range of the relation to be "Text".

Would that cover what you need?

If there are relations configured like real extension relations it should be easy to handle them in text form in #1063

**Comment by ValWood** [comment]
Yeah ignore my last comment. I think its all dealt with. I'll do some testing today when I prepare my presentation for tomorrow. I'm doing Canto at the group meeting...

**Comment by ValWood** [comment]
Can we close this one?

**Comment by kimrutherford** [comment]
There are few warning left:

failed to store extension in 558b22ef42aa6ec2: can't parse 'ABOLISHED tetermerization', in: ABOLISHED tetermerization
failed to store extension in 558b22ef42aa6ec2: can't parse 'ABOLISHED tetermerization', in: ABOLISHED tetermerization
failed to store extension in dd0dc92db12b47e2: can't parse 'during mitotic interphase', in: during mitotic interphase
failed to store extension in 6960ce3c7a8f53d0: can't parse 'AWAITING THESE SO TERMS: coincident_with(M and P silent cassettes)', in: AWAITING THESE SO TERMS: coincident_with(M and P silent cassettes), occurs_in(CL:0002674) | coincident_with(M cassette), occurs_in(CL:0002675)

but we can fix those later.

So I'll close this.
